File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/data/network/QueryNetworkTest.java
package com.kinvey.androidTest.store.data.network;

import android.content.Context;
import android.os.Message;

import androidx.test.runner.AndroidJUnit4;
import androidx.test.filters.SmallTest;
import androidx.test.platform.app.InstrumentationRegistry;

import com.google.api.client.json.GenericJson;
import com.kinvey.android.Client;
import com.kinvey.android.callback.KinveyReadCallback;
import com.kinvey.android.model.User;
import com.kinvey.android.store.DataStore;
import com.kinvey.android.store.UserStore;
import com.kinvey.androidTest.LooperThread;
import com.kinvey.androidTest.TestManager;
import com.kinvey.androidTest.callback.DefaultKinveyDeleteCallback;
import com.kinvey.androidTest.model.Location;
import com.kinvey.androidTest.model.Person;
import com.kinvey.java.Query;
import com.kinvey.java.core.KinveyClientCallback;
import com.kinvey.java.model.KinveyReadResponse;
import com.kinvey.java.query.AbstractQuery;
import com.kinvey.java.store.StoreType;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

@RunWith(AndroidJUnit4.class)
@SmallTest
public class QueryNetworkTest {

    private static final String TEST_USERNAME = "Test_UserName";
    private static final String TEST_USERNAME_2 = "Test_UserName_2";
    private static final String USERNAME = "username";
    private static final String HEIGHT = "height";
    private static final String GEOLOC = "_geoloc";
    private static final int DEFAULT_TIMEOUT = 60;

    private Client client;

    @Before
    public void setUp() throws InterruptedException, IOException {
        Context mMockContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        client = new Client.Builder(mMockContext).build();
        client.enableDebugLogging();
        final CountDownLatch latch = new CountDownLatch(1);
        LooperThread looperThread = null;
        if (!client.isUserLoggedIn()) {
            looperThread = new LooperThread(new Runnable() {
                @Override
                public void run() {
                    try {
                        UserStore.login(TestManager.USERNAME, TestManager.PASSWORD, client, new KinveyClientCallback<User>() {
                            @Override
                            public void onSuccess(User result) {
                                assertNotNull(result);
                                latch.countDown();
                            }

                            @Override
                            public void onFailure(Throwable error) {
                                assertNull(error);
                                latch.countDown();
                            }
                        });
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            });
            looperThread.start();
        } else {
            latch.countDown();
        }
        latch.await();
        if (looperThread != null) {
            looperThread.mHandler.sendMessage(new Message());
        }
    }

    private <T extends GenericJson> void clearBackend(DataStore<T> store) throws InterruptedException {
        Query query = client.query();
        query = query.notEqual("age", "100500");
        DefaultKinveyDeleteCallback deleteCallback = delete(store, query);
    }

    private <T extends GenericJson> DefaultKinveyDeleteCallback delete(final DataStore<T> store, final Query query) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyDeleteCallback callback = new DefaultKinveyDeleteCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.delete(query, callback);
            }
        });
        looperThread.start();
        latch.await(120, TimeUnit.SECONDS);
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    private DefaultKinveyReadCallback find(final DataStore<Person> store, final Query query, int seconds) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyReadCallback callback = new DefaultKinveyReadCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.find(query, callback, null);
            }
        });
        looperThread.start();
        latch.await(seconds, TimeUnit.SECONDS);
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    private DefaultKinveyReadLocCallback findLoc(final DataStore<Location> store, final Query query, int seconds) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyReadLocCallback callback = new DefaultKinveyReadLocCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.find(query, callback, null);
            }
        });
        looperThread.start();
        latch.await(seconds, TimeUnit.SECONDS);
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    private Person createPerson(String name) {
        return new Person(name);
    }

    private static class DefaultKinveyClientCallback implements KinveyClientCallback<Person> {

        private CountDownLatch latch;
        Person result;
        Throwable error;

        DefaultKinveyClientCallback(CountDownLatch latch) {
            this.latch = latch;
        }

        @Override
        public void onSuccess(Person result) {
            this.result = result;
            finish();
        }

        @Override
        public void onFailure(Throwable error) {
            this.error = error;
            finish();
        }

        void finish() {
            latch.countDown();
        }
    }

    private static class DefaultKinveyLocationCallback implements KinveyClientCallback<Location> {

        private CountDownLatch latch;
        Location result;
        Throwable error;

        DefaultKinveyLocationCallback(CountDownLatch latch) {
            this.latch = latch;
        }

        @Override
        public void onSuccess(Location result) {
            this.result = result;
            finish();
        }

        @Override
        public void onFailure(Throwable error) {
            this.error = error;
            finish();
        }

        void finish() {
            latch.countDown();
        }
    }

    private static class DefaultKinveyReadCallback implements KinveyReadCallback<Person> {

        private CountDownLatch latch;
        KinveyReadResponse<Person> result;
        Throwable error;

        DefaultKinveyReadCallback(CountDownLatch latch) {
            this.latch = latch;
        }

        @Override
        public void onSuccess(KinveyReadResponse<Person> result) {
            this.result = result;
            finish();
        }

        @Override
        public void onFailure(Throwable error) {
            this.error = error;
            finish();
        }

        void finish() {
            latch.countDown();
        }
    }

    private static class DefaultKinveyReadLocCallback implements KinveyReadCallback<Location> {

        private CountDownLatch latch;
        KinveyReadResponse<Location> result;
        Throwable error;

        DefaultKinveyReadLocCallback(CountDownLatch latch) {
            this.latch = latch;
        }

        @Override
        public void onSuccess(KinveyReadResponse<Location> result) {
            this.result = result;
            finish();
        }

        @Override
        public void onFailure(Throwable error) {
            this.error = error;
            finish();
        }

        void finish() {
            latch.countDown();
        }
    }

    private DefaultKinveyClientCallback save(final DataStore<Person> store, final Person person) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyClientCallback callback = new DefaultKinveyClientCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.save(person, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    private DefaultKinveyLocationCallback saveLoc(final DataStore<Location> store, final Location location) throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final DefaultKinveyLocationCallback callback = new DefaultKinveyLocationCallback(latch);
        LooperThread looperThread = new LooperThread(new Runnable() {
            @Override
            public void run() {
                store.save(location, callback);
            }
        });
        looperThread.start();
        latch.await();
        looperThread.mHandler.sendMessage(new Message());
        return callback;
    }

    @Test
    public void testStartsWith() throws InterruptedException {
        DataStore<Person> store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.NETWORK, client);
        clearBackend(store);
        client.getSyncManager().clear(Person.COLLECTION);
        Person person = createPerson(TEST_USERNAME);
        DefaultKinveyClientCallback saveCallback = save(store, person);
        assertNotNull(saveCallback.result);
        assertNull(saveCallback.error);
        assertNotNull(saveCallback.result.getId());
        Query query = client.query();
        query = query.startsWith(USERNAME, "Tes");
        DefaultKinveyReadCallback kinveyListCallback = find(store, query, DEFAULT_TIMEOUT);
        assertNull(kinveyListCallback.error);
        assertNotNull(kinveyListCallback.result);
        assertTrue(kinveyListCallback.result.getResult().size() > 0);
        delete(store, query);
    }

    @Test
    public void testQueryString() throws InterruptedException {
        DataStore<Person> store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.NETWORK, client);
        clearBackend(store);
        client.getSyncManager().clear(Person.COLLECTION);
        Person person = createPerson(TEST_USERNAME);
        DefaultKinveyClientCallback saveCallback = save(store, person);
        assertNotNull(saveCallback.result);
        assertNull(saveCallback.error);
        assertNotNull(saveCallback.result.getId());
        Query query = client.query();
        query = query.setQueryString("{\"username\":{\"$regex\":\"^Tes\"}}");
        DefaultKinveyReadCallback kinveyListCallback = find(store, query, DEFAULT_TIMEOUT);
        assertNull(kinveyListCallback.error);
        assertNotNull(kinveyListCallback.result);
        assertTrue(kinveyListCallback.result.getResult().size() > 0);
        delete(store, query);
    }

    @Test
    public void testAll() throws InterruptedException {
        DataStore<Location> store = DataStore.collection(Location.COLLECTION, Location.class, StoreType.NETWORK, client);
        clearBackend(store);
        client.getSyncManager().clear(Location.COLLECTION);
        Double[] geo = {3.0, 4.0};
        Location location = new Location();
        location.setGeo(geo);
        DefaultKinveyLocationCallback saveCallback = saveLoc(store, location);
        assertNotNull(saveCallback.result);
        assertNull(saveCallback.error);
        assertNotNull(saveCallback.result.getId());
        Query query = client.query();
        query = query.all(GEOLOC, geo);
        DefaultKinveyReadLocCallback kinveyListCallback = findLoc(store, query, DEFAULT_TIMEOUT);
        assertNull(kinveyListCallback.error);
        assertNotNull(kinveyListCallback.result);
        assertTrue(kinveyListCallback.result.getResult().size() > 0);
    }

    @Test
    public void testSize() throws InterruptedException {
        DataStore<Location> store = DataStore.collection(Location.COLLECTION, Location.class, StoreType.NETWORK, client);
        clearBackend(store);
        client.getSyncManager().clear(Location.COLLECTION);
        Double[] geo = {3.0, 4.0};
        Location location = new Location();
        location.setGeo(geo);
        DefaultKinveyLocationCallback saveCallback = saveLoc(store, location);
        assertNotNull(saveCallback.result);
        assertNull(saveCallback.error);
        assertNotNull(saveCallback.result.getId());
        Query query = client.query();
        query = query.size(GEOLOC, 2);
        DefaultKinveyReadLocCallback kinveyListCallback = findLoc(store, query, DEFAULT_TIMEOUT);
        assertNull(kinveyListCallback.error);
        assertNotNull(kinveyListCallback.result);
        assertTrue(kinveyListCallback.result.getResult().size() > 0);
    }

    @Test
    public void testLimit() throws InterruptedException {
        DataStore<Location> store = DataStore.collection(Location.COLLECTION, Location.class, StoreType.NETWORK, client);
        clearBackend(store);
        client.getSyncManager().clear(Location.COLLECTION);
        Double[] geo = {3.0, 4.0};
        Location location = new Location();
        location.setGeo(geo);
        DefaultKinveyLocationCallback saveCallback = saveLoc(store, location);
        assertNotNull(saveCallback.result);
        assertNull(saveCallback.error);
        assertNotNull(saveCallback.result.getId());
        saveCallback = saveLoc(store, location);
        assertNotNull(saveCallback.result);
        assertNull(saveCallback.error);
        assertNotNull(saveCallback.result.getId());
        Query query = client.query();
        query = query.setLimit(2);
        DefaultKinveyReadLocCallback kinveyListCallback = findLoc(store, query, DEFAULT_TIMEOUT);
        assertNull(kinveyListCallback.error);
        assertNotNull(kinveyListCallback.result);
        assertTrue(kinveyListCallback.result.getResult().size() == 2);
    }

    @Test
    public void testRegEx() throws InterruptedException {
        DataStore<Person> store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.NETWORK, client);
        clearBackend(store);
        client.getSyncManager().clear(Person.COLLECTION);
        Person person = createPerson(TEST_USERNAME);
        DefaultKinveyClientCallback saveCallback = save(store, person);
        assertNotNull(saveCallback.result);
        assertNull(saveCallback.error);
        assertNotNull(saveCallback.result.getId());
        Query query = client.query();
        query = query.regEx(USERNAME, "^T");
        DefaultKinveyReadCallback kinveyListCallback = find(store, query, DEFAULT_TIMEOUT);
        assertNull(kinveyListCallback.error);
        assertNotNull(kinveyListCallback.result);
        assertTrue(kinveyListCallback.result.getResult().size() > 0);
        delete(store, query);
    }

    @Test
    public void testQueryAddSort() throws InterruptedException {
        float heightPersonOne = 180;
        float heightPersonTwo = 150;
        DataStore<Person> store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.NETWORK, client);
        clearBackend(store);
        client.getSyncManager().clear(Person.COLLECTION);
        Person person = createPerson(TEST_USERNAME);
        person.setHeight(heightPersonOne);
        DefaultKinveyClientCallback saveCallback = save(store, person);
        assertNotNull(saveCallback.result);
        assertNull(saveCallback.error);
        assertNotNull(saveCallback.result.getId());
        person = createPerson(TEST_USERNAME_2);
        person.setHeight(heightPersonTwo);
        saveCallback = save(store, person);
        assertNotNull(saveCallback.result);
        assertNull(saveCallback.error);
        assertNotNull(saveCallback.result.getId());
        Query query = client.query();
        query = query.addSort(HEIGHT, AbstractQuery.SortOrder.ASC);
        DefaultKinveyReadCallback kinveyListCallback = find(store, query, DEFAULT_TIMEOUT);
        assertNull(kinveyListCallback.error);
        assertNotNull(kinveyListCallback.result);
        assertTrue(kinveyListCallback.result.getResult().size() > 1);
        assertTrue(kinveyListCallback.result.getResult().get(0).getHeight() == heightPersonTwo);
        delete(store, query);
    }

    @Test
    public void testQueryGeoWithinBox() throws InterruptedException {
        DataStore<Location> store = DataStore.collection(Location.COLLECTION, Location.class, StoreType.NETWORK, client);
        clearBackend(store);
        client.getSyncManager().clear(Location.COLLECTION);

        Double[] geo = {30.0, 40.0};
        Location location = new Location();
        location.setGeo(geo);

        DefaultKinveyLocationCallback saveCallback = saveLoc(store, location);
        assertNotNull(saveCallback.result);
        assertNull(saveCallback.error);

        Query query = client.query();
        Query geoQuery = query.withinBox(GEOLOC, 20, 20, 40, 50);
        DefaultKinveyReadLocCallback kinveyListCallback = findLoc(store, geoQuery, DEFAULT_TIMEOUT);
        assertNull(kinveyListCallback.error);
        assertNotNull(kinveyListCallback.result);
        assertTrue(kinveyListCallback.result.getResult().size() > 0);
    }

    @Test
    public void testQueryGeoNearSphere() throws InterruptedException {
        DataStore<Location> store = DataStore.collection(Location.COLLECTION, Location.class, StoreType.NETWORK, client);
        clearBackend(store);
        client.getSyncManager().clear(Location.COLLECTION);

        Double[] geo = {30.0, 40.0};
        Location location = new Location();
        location.setGeo(geo);

        DefaultKinveyLocationCallback saveCallback = saveLoc(store, location);
        assertNotNull(saveCallback.result);
        assertNull(saveCallback.error);

        Query query1 = client.query();
        Query geoQueryNearSphere = query1.nearSphere(GEOLOC, 20, 50);
        Query query2 = client.query();
        Query geoQueryNearSphereDistance = query2.nearSphere(GEOLOC, 20, 50, 100000);

        DefaultKinveyReadLocCallback kinveyFindCallback1 = findLoc(store, geoQueryNearSphere, DEFAULT_TIMEOUT);
        DefaultKinveyReadLocCallback kinveyFindCallback2 = findLoc(store, geoQueryNearSphereDistance, DEFAULT_TIMEOUT);

        assertNull(kinveyFindCallback1.error);
        assertNotNull(kinveyFindCallback1.result);
        assertTrue(kinveyFindCallback1.result.getResult().size() > 0);

        assertNull(kinveyFindCallback2.error);
        assertNotNull(kinveyFindCallback2.result);
        assertTrue(kinveyFindCallback2.result.getResult().size() > 0);
    }

    @Test
    public void testQueryGeoWithinPolygon() throws InterruptedException {
        DataStore<Location> store = DataStore.collection(Location.COLLECTION, Location.class, StoreType.NETWORK, client);
        clearBackend(store);
        client.getSyncManager().clear(Location.COLLECTION);

        Double[] geo = {30.0, 40.0};
        Location location = new Location();
        location.setGeo(geo);

        DefaultKinveyLocationCallback saveCallback = saveLoc(store, location);
        assertNotNull(saveCallback.result);
        assertNull(saveCallback.error);

        Query query1 = client.query();
        Query geoQueryWithinPolygon = query1.withinPolygon(GEOLOC,
                30, 20,
                40, 30,
                50, 40,
                60, 50);

        DefaultKinveyReadLocCallback kinveyFindCallback = findLoc(store, geoQueryWithinPolygon, DEFAULT_TIMEOUT);

        assertNull(kinveyFindCallback.error);
        assertNotNull(kinveyFindCallback.result);
        assertTrue(kinveyFindCallback.result.getResult().size() > 0);
    }

    @After
    public void tearDown() {
        client.performLockDown();
        if (client.getKinveyHandlerThread() != null) {
            try {
                client.stopKinveyHandlerThread();
            } catch (Throwable throwable) {
                throwable.printStackTrace();
            }
        }
    }
}

File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/data/network/QueryNetworkTest.kt
package com.kinvey.androidTest.store.data.network

import android.os.Message
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.SmallTest
import androidx.test.platform.app.InstrumentationRegistry
import com.google.api.client.json.GenericJson
import com.kinvey.android.Client
import com.kinvey.android.Client.Builder
import com.kinvey.android.callback.KinveyReadCallback
import com.kinvey.android.model.User
import com.kinvey.android.store.DataStore
import com.kinvey.android.store.DataStore.Companion.collection
import com.kinvey.android.store.UserStore.Companion.login
import com.kinvey.androidTest.LooperThread
import com.kinvey.androidTest.TestManager
import com.kinvey.androidTest.callback.DefaultKinveyDeleteCallback
import com.kinvey.androidTest.model.Location
import com.kinvey.androidTest.model.Person
import com.kinvey.java.AbstractClient
import com.kinvey.java.Query
import com.kinvey.java.core.KinveyClientCallback
import com.kinvey.java.model.KinveyReadResponse
import com.kinvey.java.query.AbstractQuery.SortOrder
import com.kinvey.java.store.StoreType
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.io.IOException
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

@RunWith(AndroidJUnit4::class)
@SmallTest
class QueryNetworkTest {

    private var client: Client<*>? = null

    @Before
    @Throws(InterruptedException::class, IOException::class)
    fun setUp() {
        val mMockContext = InstrumentationRegistry.getInstrumentation().targetContext
        client = Builder<User>(mMockContext).build()
        client?.enableDebugLogging()
        val latch = CountDownLatch(1)
        var looperThread: LooperThread? = null
        if (client?.isUserLoggedIn == false) {
            looperThread = LooperThread(Runnable {
                try {
                    login<User>(TestManager.USERNAME, TestManager.PASSWORD, client as AbstractClient<User>,
                    object : KinveyClientCallback<User> {
                        override fun onSuccess(result: User?) {
                            assertNotNull(result)
                            latch.countDown()
                        }

                        override fun onFailure(error: Throwable?) {
                            assertNull(error)
                            latch.countDown()
                        }
                    })
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            })
            looperThread.start()
        } else {
            latch.countDown()
        }
        latch.await()
        looperThread?.mHandler?.sendMessage(Message())
    }

    @Throws(InterruptedException::class)
    private fun <T : GenericJson> clearBackend(store: DataStore<T>) {
        var query = client?.query() as Query
        query = query?.notEqual("age", "100500")
        val deleteCallback = delete(store, query)
    }

    @Throws(InterruptedException::class)
    private fun <T : GenericJson> delete(store: DataStore<T>, query: Query): DefaultKinveyDeleteCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyDeleteCallback(latch)
        val looperThread = LooperThread(Runnable { store.delete(query, callback) })
        looperThread.start()
        latch.await(120, TimeUnit.SECONDS)
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    private fun find(store: DataStore<Person>, query: Query, seconds: Int): DefaultKinveyReadCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyReadCallback(latch)
        val looperThread = LooperThread(Runnable { store?.find(query, callback, null) })
        looperThread.start()
        latch.await(seconds.toLong(), TimeUnit.SECONDS)
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    private fun findLoc(store: DataStore<Location>, query: Query, seconds: Int): DefaultKinveyReadLocCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyReadLocCallback(latch)
        val looperThread = LooperThread(Runnable { store.find(query, callback, null) })
        looperThread.start()
        latch.await(seconds.toLong(), TimeUnit.SECONDS)
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    private fun createPerson(name: String): Person {
        return Person(name)
    }

    private class DefaultKinveyClientCallback(val latch: CountDownLatch) : KinveyClientCallback<Person> {
        var result: Person? = null
        var error: Throwable? = null
        override fun onSuccess(result: Person?) {
            this.result = result
            finish()
        }
        override fun onFailure(error: Throwable?) {
            this.error = error
            finish()
        }
        fun finish() {
            latch.countDown()
        }
    }

    private class DefaultKinveyLocationCallback(val latch: CountDownLatch) : KinveyClientCallback<Location> {
        var result: Location? = null
        var error: Throwable? = null
        override fun onSuccess(result: Location?) {
            this.result = result
            finish()
        }
        override fun onFailure(error: Throwable?) {
            this.error = error
            finish()
        }
        fun finish() {
            latch.countDown()
        }
    }

    private class DefaultKinveyReadCallback (val latch: CountDownLatch) : KinveyReadCallback<Person> {
        var result: KinveyReadResponse<Person>? = null
        var error: Throwable? = null
        override fun onSuccess(result: KinveyReadResponse<Person>?) {
            this.result = result
            finish()
        }
        override fun onFailure(error: Throwable?) {
            this.error = error
            finish()
        }
        fun finish() {
            latch.countDown()
        }
    }

    private class DefaultKinveyReadLocCallback(val latch: CountDownLatch) : KinveyReadCallback<Location> {
        var result: KinveyReadResponse<Location>? = null
        var error: Throwable? = null
        override fun onSuccess(result: KinveyReadResponse<Location>?) {
            this.result = result
            finish()
        }
        override fun onFailure(error: Throwable?) {
            this.error = error
            finish()
        }
        fun finish() {
            latch.countDown()
        }
    }

    @Throws(InterruptedException::class)
    private fun save(store: DataStore<Person>, person: Person): DefaultKinveyClientCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyClientCallback(latch)
        val looperThread = LooperThread(Runnable { store?.save(person, callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Throws(InterruptedException::class)
    private fun saveLoc(store: DataStore<Location>, location: Location): DefaultKinveyLocationCallback {
        val latch = CountDownLatch(1)
        val callback = DefaultKinveyLocationCallback(latch)
        val looperThread = LooperThread(Runnable { store.save(location, callback) })
        looperThread.start()
        latch.await()
        looperThread.mHandler?.sendMessage(Message())
        return callback
    }

    @Test
    @Throws(InterruptedException::class)
    fun testStartsWith() {
        val store = collection(Person.COLLECTION, Person::class.java, StoreType.NETWORK, client)
        clearBackend(store)
        client?.syncManager?.clear(Person.COLLECTION)
        val person = createPerson(TEST_USERNAME)
        val saveCallback = save(store, person)
        assertNotNull(saveCallback.result)
        assertNull(saveCallback.error)
        assertNotNull(saveCallback.result?.id)
        var query = client?.query()
        query = query?.startsWith(USERNAME, "Tes") as Query
        val kinveyListCallback = find(store, query, DEFAULT_TIMEOUT)
        assertNull(kinveyListCallback.error)
        assertNotNull(kinveyListCallback.result)
        assertTrue(kinveyListCallback.result?.result?.size ?: 0 > 0)
        delete(store, query)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testQueryString() {
        val store = collection(Person.COLLECTION, Person::class.java, StoreType.NETWORK, client)
        clearBackend(store)
        client?.syncManager?.clear(Person.COLLECTION)
        val person = createPerson(TEST_USERNAME)
        val saveCallback = save(store, person)
        assertNotNull(saveCallback.result)
        assertNull(saveCallback.error)
        assertNotNull(saveCallback.result?.id)
        var query = client?.query()
        query = query?.setQueryString("{\"username\":{\"\$regex\":\"^Tes\"}}") as Query
        val kinveyListCallback = find(store, query, DEFAULT_TIMEOUT)
        assertNull(kinveyListCallback.error)
        assertNotNull(kinveyListCallback.result)
        assertTrue(kinveyListCallback.result?.result?.size ?: 0 > 0)
        delete(store, query)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testAll() {
        val store = collection(Location.COLLECTION, Location::class.java, StoreType.NETWORK, client)
        clearBackend(store)
        client?.syncManager?.clear(Location.COLLECTION)
        val geo = arrayOf<Double>(3.0, 4.0)
        val location = Location()
        location.geo = geo
        val saveCallback = saveLoc(store, location)
        assertNotNull(saveCallback.result)
        assertNull(saveCallback.error)
        assertNotNull(saveCallback.result?.id)
        var query = client?.query()
        query = query?.all(GEOLOC, geo) as Query
        val kinveyListCallback = findLoc(store, query, DEFAULT_TIMEOUT)
        assertNull(kinveyListCallback.error)
        assertNotNull(kinveyListCallback.result)
        assertTrue(kinveyListCallback.result?.result?.size ?: 0 > 0)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testSize() {
        val store = collection(Location.COLLECTION, Location::class.java, StoreType.NETWORK, client)
        clearBackend(store)
        client?.syncManager?.clear(Location.COLLECTION)
        val geo = arrayOf(3.0, 4.0)
        val location = Location()
        location.geo = geo
        val saveCallback = saveLoc(store, location)
        assertNotNull(saveCallback.result)
        assertNull(saveCallback.error)
        assertNotNull(saveCallback.result?.id)
        var query = client?.query()
        query = query?.size(GEOLOC, 2) as Query
        val kinveyListCallback = findLoc(store, query, DEFAULT_TIMEOUT)
        assertNull(kinveyListCallback.error)
        assertNotNull(kinveyListCallback.result)
        assertTrue(kinveyListCallback.result?.result?.size ?: 0 > 0)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testLimit() {
        val store = collection(Location.COLLECTION, Location::class.java, StoreType.NETWORK, client)
        clearBackend(store)
        client?.syncManager?.clear(Location.COLLECTION)
        val geo = arrayOf(3.0, 4.0)
        val location = Location()
        location.geo = geo
        var saveCallback = saveLoc(store, location)
        assertNotNull(saveCallback.result)
        assertNull(saveCallback.error)
        assertNotNull(saveCallback.result?.id)
        saveCallback = saveLoc(store, location)
        assertNotNull(saveCallback.result)
        assertNull(saveCallback.error)
        assertNotNull(saveCallback.result?.id)
        var query = client?.query()
        query = query?.setLimit(2) as Query
        val kinveyListCallback = findLoc(store, query, DEFAULT_TIMEOUT)
        assertNull(kinveyListCallback.error)
        assertNotNull(kinveyListCallback.result)
        assertTrue(kinveyListCallback.result?.result?.size == 2)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testRegEx() {
        val store = collection(Person.COLLECTION, Person::class.java, StoreType.NETWORK, client)
        clearBackend(store)
        client?.syncManager?.clear(Person.COLLECTION)
        val person = createPerson(TEST_USERNAME)
        val saveCallback = save(store, person)
        assertNotNull(saveCallback.result)
        assertNull(saveCallback.error)
        assertNotNull(saveCallback.result?.id)
        var query = client?.query() as Query
        query = query.regEx(USERNAME, "^T")
        val kinveyListCallback = find(store, query, DEFAULT_TIMEOUT)
        assertNull(kinveyListCallback.error)
        assertNotNull(kinveyListCallback.result)
        assertTrue(kinveyListCallback.result?.result?.size ?: 0 > 0)
        delete(store, query)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testQueryAddSort() {
        val heightPersonOne = 180f
        val heightPersonTwo = 150f
        val store = collection(Person.COLLECTION, Person::class.java, StoreType.NETWORK, client)
        clearBackend(store)
        client?.syncManager?.clear(Person.COLLECTION)
        var person = createPerson(TEST_USERNAME)
        person.height = heightPersonOne
        var saveCallback = save(store, person)
        assertNotNull(saveCallback.result)
        assertNull(saveCallback.error)
        assertNotNull(saveCallback.result?.id)
        person = createPerson(TEST_USERNAME_2)
        person.height = heightPersonTwo
        saveCallback = save(store, person)
        assertNotNull(saveCallback.result)
        assertNull(saveCallback.error)
        assertNotNull(saveCallback.result?.id)
        var query = client?.query() as Query
        query = query?.addSort(HEIGHT, SortOrder.ASC)
        val kinveyListCallback = find(store, query, DEFAULT_TIMEOUT)
        assertNull(kinveyListCallback.error)
        assertNotNull(kinveyListCallback.result)
        assertTrue(kinveyListCallback.result?.result?.size ?: 0 > 1)
        assertTrue(kinveyListCallback.result?.result?.get(0)?.height == heightPersonTwo)
        delete(store, query)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testQueryGeoWithinBox() {
        val store = collection(Location.COLLECTION, Location::class.java, StoreType.NETWORK, client)
        clearBackend(store)
        client?.syncManager?.clear(Location.COLLECTION)
        val geo = arrayOf(30.0, 40.0)
        val location = Location()
        location.geo = geo
        val saveCallback = saveLoc(store, location)
        assertNotNull(saveCallback.result)
        assertNull(saveCallback.error)
        val query = client?.query()
        val geoQuery = query?.withinBox(GEOLOC, 20.0, 20.0, 40.0, 50.0) as Query
        val kinveyListCallback = findLoc(store, geoQuery, DEFAULT_TIMEOUT)
        assertNull(kinveyListCallback.error)
        assertNotNull(kinveyListCallback.result)
        assertTrue(kinveyListCallback.result?.result?.size ?: 0 > 0)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testQueryGeoNearSphere() {
        val store = collection(Location.COLLECTION, Location::class.java, StoreType.NETWORK, client)
        clearBackend(store)
        client?.syncManager?.clear(Location.COLLECTION)
        val geo = arrayOf(30.0, 40.0)
        val location = Location()
        location.geo = geo
        val saveCallback = saveLoc(store, location)
        assertNotNull(saveCallback.result)
        assertNull(saveCallback.error)
        val query1 = client?.query()
        val geoQueryNearSphere = query1?.nearSphere(GEOLOC, 20.0, 50.0) as Query
        val query2 = client?.query()
        val geoQueryNearSphereDistance = query2?.nearSphere(GEOLOC, 20.0, 50.0, 100000.0) as Query
        val kinveyFindCallback1 = findLoc(store, geoQueryNearSphere, DEFAULT_TIMEOUT)
        val kinveyFindCallback2 = findLoc(store, geoQueryNearSphereDistance, DEFAULT_TIMEOUT)
        assertNull(kinveyFindCallback1.error)
        assertNotNull(kinveyFindCallback1.result)
        assertTrue(kinveyFindCallback1.result?.result?.size ?: 0 > 0)
        assertNull(kinveyFindCallback2.error)
        assertNotNull(kinveyFindCallback2.result)
        assertTrue(kinveyFindCallback2.result?.result?.size ?: 0 > 0)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testQueryGeoWithinPolygon() {
        val store = collection(Location.COLLECTION, Location::class.java, StoreType.NETWORK, client)
        clearBackend(store)
        client!!.syncManager.clear(Location.COLLECTION)
        val geo = arrayOf(30.0, 40.0)
        val location = Location()
        location.geo = geo
        val saveCallback = saveLoc(store, location)
        assertNotNull(saveCallback.result)
        assertNull(saveCallback.error)
        val query1 = client?.query()
        val geoQueryWithinPolygon = query1?.withinPolygon(GEOLOC,
                30.0, 20.0, 40.0, 30.0,
                50.0, 40.0, 60.0, 50.0) as Query
        val kinveyFindCallback = findLoc(store, geoQueryWithinPolygon, DEFAULT_TIMEOUT)
        assertNull(kinveyFindCallback.error)
        assertNotNull(kinveyFindCallback.result)
        assertTrue(kinveyFindCallback.result?.result?.size ?: 0 > 0)
    }

    @After
    fun tearDown() {
        client?.performLockDown()
        if (Client.kinveyHandlerThread != null) {
            try {
                client?.stopKinveyHandlerThread()
            } catch (throwable: Throwable) {
                throwable.printStackTrace()
            }
        }
    }

    companion object {
        private const val TEST_USERNAME = "Test_UserName"
        private const val TEST_USERNAME_2 = "Test_UserName_2"
        private const val USERNAME = "username"
        private const val HEIGHT = "height"
        private const val GEOLOC = "_geoloc"
        private const val DEFAULT_TIMEOUT = 60
    }
}